lib/Sema/TypeCheckStmt.cpp

```diff
@@ -1246,8 +1246,21 @@ Stmt *StmtChecker::visitBraceStmt(BraceStmt *BS) {
 
       bool hadTypeError = TC.typeCheckExpression(SubExpr, DC, TypeLoc(),
                                                  CTP_Unused, options);
-      if (isDiscarded && !hadTypeError)
+
+      // If a closure expression is unused, the user might have intended
+      // to write "do { ... }".
+      auto *CE = dyn_cast<ClosureExpr>(SubExpr);
```

Comment:
I noticed that I'm not actually checking `isDiscarded` here. Surprised this didn't break some tests. Will fix this evening.